new feature(CmsKit): Marked Item

suhaib-mousa opened the pull request:
Resolves #4763 

Add a marked item feature to mark items as favorite, starred, flagged, or bookmarked.

- Added Marked Toggling Component to mark items via pre-defined icons/emojis.
- Created a Marked Item widget for toggling marks on items.

### Checklist

- [X] I fully tested it as developer / designer and created unit / integration tests
- [X] I documented it.

## Future Work

* Build a page for the marked items for the user.

suhaib-mousa committed:
new feature(CmsKit): Marked Item

Add Marked Item feature to allow marking items as  favorite, starred, flagged, or bookmarked.

- Added Marked Toggling Component to mark items via pre-defined icons/emojis.
- Created Marked Item widget for toggling marks on items.

Documentation: Provided detailed usage and configuration instructions in the CMS Kit Module documentation.

OthmanMaher-Dimensions commented:
@maliming 
Can you review this request please ?

enisn commented on modules/cms-kit/src/Volo.CmsKit.Domain/Volo/CmsKit/MarkedItems/CmsKitMarkedItemOptions.cs:
```diff
@@ -0,0 +1,10 @@
+public class CmsKitMarkedItemOptions
+{
+    [NotNull]
+    public List<MarkedItemEntityTypeDefinition> EntityTypes { get; } = new();
```
What if I define multiple definitions for the same EntityType? Since it's queried by `Single()` it throws an exception but a more clear exception can be thrown

enisn commented on modules/cms-kit/src/Volo.CmsKit.Domain/Volo/CmsKit/MarkedItems/MarkedItemManager.cs:
```diff
@@ -0,0 +1,58 @@
+                )
+            );
+        return true;
+    }
```
Not sure about the name of the operation, can it be a better name that defines what it does?

enisn commented on modules/cms-kit/src/Volo.CmsKit.Domain/Volo/CmsKit/MarkedItems/MarkedItemManager.cs:
```diff
@@ -0,0 +1,58 @@
+        return true;
+    }
+
+    public virtual async Task<MarkedItemEntityTypeDefinition> GetAsync(
+        [NotNull] string entityType)
+    {
+        Check.NotNullOrWhiteSpace(entityType, nameof(entityType));
+
+        return await MarkedItemDefinitionStore.GetAsync(entityType);
+    }
```
This method is not necessary, `MarkedItemDefinitionStore.GetAsync` already checks NullOrEmpty, after removing this check from here, this method has no more purpose. We can use directly `MarkedItemDefinitionStore` in the AppService

enisn commented on modules/cms-kit/src/Volo.CmsKit.Domain/Volo/CmsKit/MarkedItems/IUserMarkedItemRepository.cs:
```diff
@@ -0,0 +1,31 @@
+        CancellationToken cancellationToken = default
+    );
+
+    Task<IQueryable<UserMarkedItem>> GetQueryForUserAsync(
```
Please do not expose `IQueryable` from repositories in the modules. Since I reviewed, this is not even used but tests, we can easily remove it

suhaib-mousa replied:
![image](https://github.com/user-attachments/assets/3e95e119-c3bc-464b-b01d-7e5b31dbf99c)
I mentioned the need for the IQueryable method in the documentation... So, if it is better to remove it, we will remove it.

enisn replied:
Can you go with the same approach with tags?
https://github.com/abpframework/abp/blob/583f3c5b3cd80bb9820209f1b1bae611f840c011/modules/cms-kit/src/Volo.CmsKit.Domain/Volo/CmsKit/Tags/IEntityTagRepository.cs#L25

enisn commented on modules/cms-kit/src/Volo.CmsKit.Public.Application.Contracts/Volo/CmsKit/Public/MarkedItems/IMarkedItemPublicAppService.cs:
```diff
@@ -0,0 +1,10 @@
+
+public interface IMarkedItemPublicAppService : IApplicationService
+{
+    Task<MarkedItemWithToggleDto> GetForToggleAsync(string entityType, string entityId);
```
The name `GetForToggleAsync` is not good enough. How about making it simple like `GetAsync` or `GetForUserAsync`?

enisn commented on modules/cms-kit/src/Volo.CmsKit.Public.Application/Volo/CmsKit/Public/MarkedItems/MarkedItemPublicAppService.cs:
```diff
@@ -0,0 +1,69 @@
+using Volo.Abp.GlobalFeatures;
+using Volo.Abp.Users;
+using Volo.CmsKit.GlobalFeatures;
+using Volo.CmsKit.MarkedItems;
+
+namespace Volo.CmsKit.Public.MarkedItems;
+
```
```suggestion
using Volo.CmsKit.MarkedItems;
using Volo.CmsKit.Features;
using Volo.Abp.Features;

namespace Volo.CmsKit.Public.MarkedItems;

[RequiresFeature(CmsKitFeatures.PageEnable)]
```

enisn replied:
```suggestion
[RequiresFeature(CmsKitFeatures.PageEnable)]
```

enisn commented on modules/cms-kit/src/Volo.CmsKit.Public.Application/Volo/CmsKit/Public/MarkedItems/MarkedItemPublicAppService.cs:
```diff
@@ -0,0 +1,69 @@
+using Volo.Abp.Users;
+using Volo.CmsKit.GlobalFeatures;
+using Volo.CmsKit.MarkedItems;
+
```
```suggestion
using Volo.CmsKit.Features;
using Volo.Abp.Features;

```

enisn committed:
Update MarkedItemManager.cs

enisn committed:
Call directly FirstOrDefaultAsync

enisn committed:
Make injected properties protected

enisn committed:
Remove empty line

suhaib-mousa committed:
Update modules/cms-kit/src/Volo.CmsKit.Public.Application/Volo/CmsKit/Public/MarkedItems/MarkedItemPublicAppService.cs

Co-authored-by: Enis Necipoglu <info@enisnecipoglu.com>

suhaib-mousa committed:
Update modules/cms-kit/src/Volo.CmsKit.Public.Application/Volo/CmsKit/Public/MarkedItems/MarkedItemPublicAppService.cs

Co-authored-by: Enis Necipoglu <info@enisnecipoglu.com>

suhaib-mousa committed:
Merge branch 'abpframework:dev' into features/marked-item-feature

suhaib-mousa committed:
Improve MarkedItems module based on review feedback

- Refactor CmsKitMarkedItemOptions to prevent duplicate entity type definitions
- Enhance exception handling in DefaultMarkedItemDefinitionStore
- Remove redundant GetAsync method from MarkedItemManager
- Remove IQueryable exposure from IUserMarkedItemRepository
- Align EntityTagRepository approach with other repositories
- Rename GetForToggleAsync to GetAsync or GetForUserAsync in IMarkedItemPublicAppService

These changes improve code clarity, consistency, and adherence to ABP.io best practices.

suhaib-mousa commented:
@enisn 

I've made the changes based on your feedback. 

Could you please review these changes and let me know if any further modifications are needed? Thank you for your thorough review and suggestions.

suhaib-mousa committed:
resolve the feature checking to check on markedItem feature

suhaib-mousa commented:
@enisn @hikalkan 
Hi, any updates?

enisn commented:
Hi, thanks for your contribution, We'll check it again soon

suhaib-mousa commented:
@enisn @berkansasmaz 
Can you please clarify the reason for the failed test? 😅
